src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -753,6 +758,81 @@ protected Class<?> findClass(String moduleName, String name) {
         return null;
     }
 
+    private enum LoadedState { CLAIMED, LOADING, CCE };
+
+    private final LoadedState getPlaceholder(String name) {
+        Thread thread = Thread.currentThread();
+        Thread pt = placeholders.putIfAbsent(name, thread);
+        if (pt == null) {
+            return LoadedState.CLAIMED;
+        } else if (pt == thread) {
+            return LoadedState.CCE;
+        } else {
+            // Another thread is loading this class.
+            return LoadedState.LOADING;
+        }
+    }
+
+    private final void removePlaceholder(String name) {
+        placeholders.remove(name);
+        // Notify threads waiting on the class loader lock
+        // that this class has been loaded or failed.
+        notifyAll();
+    }
+
+    // We only get here if the application has released the
+    // classloader lock when another thread was in the middle of loading a
+    // superclass/superinterface for this class, and now
+    // this thread is also trying to load this class.
+    // To minimize surprises, this thread waits while the first thread
+    // that started to load a class completes the loading or fails.
+    private LoadedState waitForPlaceholder(String name) {
+        LoadedState state;
+        int count = 0;
+        while ((state = getPlaceholder(name)) == LoadedState.LOADING) {
+            try {
+                wait();
+            } catch (InterruptedException e) {
+                // keep waiting, must be uninterruptible
+            }
```

Comment:
This will swallow the interrupt so I assume you'l need a boolean to record that the wait was interrupt and then re-assert the interrupt status before exiting, as in `if (interrupted) Thread.currentThread().interrupt();`

Reply:
Hi Alan, thank you for looking at this.  Is there an example of other code that does this so I can see what I'm supposed to do?